keps/sig-apps/3939-include-terminating-pods-as-active/README.md

```diff
@@ -0,0 +1,811 @@
+ // as active. If the field is true, then the Deployment controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+In [Option for acknowledging terminating Pods in Deployment rolling update](https://github.com/kubernetes/kubernetes/issues/107920)
+there was a request to add this as part of the `DeploymentStrategy` field.  Generally, handling terminating pods as active can be useful in both RollingUpdates and Recreating rollouts.  Having this field for both strategies allows for handling of terminating pods in both cases.  
+
+Deployments create ReplicaSets so there is a need to add a field in the ReplicaSet as well.  Since ReplicaSets are not typically
+set by users, we should add a field to the ReplicaSet that is set from the DeploymentSpec.  
+
+```golang
+// ReplicaSetSpec is the specification of a ReplicaSet.
+// As the internal representation of a ReplicaSet, it must have
+// a Template set.
+type ReplicaSetSpec struct {
+  ...
+ // TerminatingAsActive specifies if the Deployments should include terminating pods
+ // as active. If the field is true, then the Deployment controller will include active pods
+ // to mean running or terminating pods
+ // +optional
+ TerminatingAsActive *bool
+}
+```
+
+### Implementation
+
+Generally, both the Job controller and ReplicaSets utilize `FilterActivePods` in their reconciliation loop.  `FilterActivePods` gets a list of pods that are not terminating.  This KEP will include terminating pods in this list.
+
+```golang
+// FilterActivePods returns pods that have not terminated.
+func FilterActivePods(pods []*v1.Pod, terminatingPods bool) []*v1.Pod {
+ var result []*v1.Pod
+ for _, p := range pods {
+  if IsPodActive(p) {
+   result = append(result, p)
+  } else if IsPodTerminating(p) && terminatingPods {
+      result = append(result, p)
+  } else {
+   klog.V(4).Infof("Ignoring inactive pod %v/%v in state %v, deletion time %v",
+    p.Namespace, p.Name, p.Status.Phase, p.DeletionTimestamp)
+  }
+ }
+ return result
+}
+
+func IsPodTerminating(p *v1.Pod) bool {
+ return v1.PodSucceeded != p.Status.Phase &&
+  v1.PodFailed != p.Status.Phase &&
+  p.DeletionTimestamp != nil
+}
+```
+
+The Job Controller uses this list to determine if there is a mismatch of active pods between expected values in the JobSpec.  
+Including active pods in this list allows the job controller to wait until these terminating pods.
+
+[Filter Active Pods Usage in Job Controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go#L749) filters the active pods.
+
+For the Deployment/ReplicaSet, ReplicaSets [filter out active pods](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/replicaset/replica_set.go#L692).  The implementation for this should include reading the deployment field and setting the replicaset the same field in the replicaset.  
```

Comment:
As mentioned above, I am not convinced we should do this as there are more things to consider for replica sets and deployments.